src/inmanta/execute/runtime.py

```diff
@@ -527,6 +537,18 @@ def get_value(self) -> "Instance":
     def __str__(self) -> str:
         return "OptionVariable %s %s = %s" % (self.myself, self.attribute, self.value)
 
+    def get_progress_potential(self) -> int:
+        """How many are actually waiting for us"""
+        result = len(self.waiters)
+
+        from inmanta.ast.attribute import RelationAttribute
+
+        if isinstance(self.attribute, RelationAttribute):
+            # Ensure that relationships with a type hint cannot end up in the zerowaiters queue
+            # of the scheduler. We know the order in which those types can be frozen safely.
+            result += int(self.attribute.has_type_hint())
```

Comment:
I would prefer to add `has_type_hint` to all `Attribute` types and have it return False by default. (avoid instanceof)